rootex/core/renderer/shaders/god_rays_pixel_shader.hlsl

```diff
@@ -0,0 +1,63 @@
+#define THRESHOLD 0.95f
+
+cbuffer cbGodRays : register(b0)
+{
+	float3 sunScreenSpacePos;
+	int numSamples;
+	float density;
+	float weight;
+	float decay;
+	float exposure;
+	float2 screenDims;
+};
+
+SamplerState SampleType : register(s0);
+
+Texture2D InputTexture : register(t0);
+Texture2D<uint2> StencilTexture : register(t1);
+
+struct GodRaysVSOutput
+{
+	float4 pos : SV_POSITION;
+	float2 tex : TEXCOORD;
+};
+
+float4 SelectiveSample(float2 texCoord)
+{
+	uint3 index = uint3(int2(screenDims * texCoord), 0);
+	uint stencil = StencilTexture.Load(index).g;
+
+	return InputTexture.Sample(SampleType, texCoord) * (1 - stencil);
+}
+
+float4 main(GodRaysVSOutput input) : SV_TARGET
+{
+	float2 texCoord = input.tex.xy;
+
+	float2 deltaTexCoord = input.tex - sunScreenSpacePos.xy;
+	deltaTexCoord *= (1.0f / float(numSamples)) * density;
+
+	// float3 color = smoothstep(THRESHOLD, 1.0f, SelectiveSample(texCoord).xyz);
+	float3 color = SelectiveSample(texCoord).xyz;
+
+	float illumination = 1.0f;
+
+	for (int i = 0; i < numSamples; i++)
```

Comment:
this is not a comment but can you try profling the draw for this code versus, defining `numSamples` as a constant in this file.
Loops are often not used in shaders because they can cause branching leading to GPU cores being stalled. If they're used then the loop condition is made to be deterministic at compile time. But I wonder what happens when we use a constant buffer in here...